### PR TITLE
spark-on-k8s sensor - add driver logs

### PR DESCRIPTION
**Pull request by bbenzikry**
Simple implementation for appending driver pod logs on completion when using SparkKubernetesSensor. 
Originally we discussed monitoring the pod as part of the operator, similar to the implementation for the [k8s pod launcher](https://github.com/apache/airflow/blob/c12e33efa99357cfd017cecf4a4852c7b2961a1d/airflow/kubernetes/pod_launcher.py#L127) but decided against it for now so we don't introduce breaking changes.

### CODE REVIEW

**Comment by mik-laj** [comment]
@roitvt Can I ask for a review?

**Comment by roitvt** [comment]
> @roitvt Can I ask for a review?

For sure I'll do it tomorrow :)

**Comment by bbenzikry** [comment]
@roitvt, sorry to bother you - do you think you'll have time to take a look this week?

**Comment by roitvt** [comment]
> @roitvt, sorry to bother you - do you think you'll have time to take a look this week?

Hi Beni, I'm really sorry I'll do it this week. thank you very much for adding this functionality and I'll be glad to have a talk with you about using this integration and Spark on K8s :)

**Comment by bbenzikry** [comment]
> > @roitvt, sorry to bother you - do you think you'll have time to take a look this week?
> 
> Hi Beni, I'm really sorry I'll do it this week. thank you very much for adding this functionality and I'll be glad to have a talk with you about using this integration and Spark on K8s :)

Great, thanks 👍 
I'll be happy to discuss, I'm also interested in some design considerations you had when writing the operator.

**Comment by roitvt** [comment]
LGTM works well with spark-pi.
Great addition thanks :) 
@bbenzikry @mik-laj

**Comment by roitvt** [comment]
> > > @roitvt, sorry to bother you - do you think you'll have time to take a look this week?
> > 
> > 
> > Hi Beni, I'm really sorry I'll do it this week. thank you very much for adding this functionality and I'll be glad to have a talk with you about using this integration and Spark on K8s :)
> 
> Great, thanks 👍
> I'll be happy to discuss, I'm also interested in some design considerations you had when writing the operator.



> > > @roitvt, sorry to bother you - do you think you'll have time to take a look this week?
> > 
> > 
> > Hi Beni, I'm really sorry I'll do it this week. thank you very much for adding this functionality and I'll be glad to have a talk with you about using this integration and Spark on K8s :)
> 
> Great, thanks 👍
> I'll be happy to discuss, I'm also interested in some design considerations you had when writing the operator.



> > @roitvt, sorry to bother you - do you think you'll have time to take a look this week?
> 
> Hi Beni, I'm really sorry I'll do it this week. thank you very much for adding this functionality and I'll be glad to have a talk with you about using this integration and Spark on K8s :)
What is your mail?

**Comment by roitvt** [comment]
> > > @roitvt, sorry to bother you - do you think you'll have time to take a look this week?
> > 
> > 
> > Hi Beni, I'm really sorry I'll do it this week. thank you very much for adding this functionality and I'll be glad to have a talk with you about using this integration and Spark on K8s :)
> 
> Great, thanks 👍
> I'll be happy to discuss, I'm also interested in some design considerations you had when writing the operator.

What is your mail?

**Review by roitvt — APPROVED** [review]

**Comment by bbenzikry** [comment]
> > > > @roitvt, sorry to bother you - do you think you'll have time to take a look this week?
> > > 
> > > 
> > > Hi Beni, I'm really sorry I'll do it this week. thank you very much for adding this functionality and I'll be glad to have a talk with you about using this integration and Spark on K8s :)
> > 
> > 
> > Great, thanks 👍
> > I'll be happy to discuss, I'm also interested in some design considerations you had when writing the operator.
> 
> What is your mail?

sent on FB

**Comment by bbenzikry** [comment]
@mik-laj Can this be merged?

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py [hunk 1]
```diff
@@ -99,19 +97,76 @@ def create_custom_resource_definition(self,
...
+        :type version: str
+        :param namespace: kubernetes namespace
+        :type namespace: str
+        """
+
+        api = client.CoreV1Api(self.get_conn())
+        watcher = watch.Watch()
+        return (
+            watcher,
+            watcher.stream(
```
Comment (mik-laj) [inline airflow/providers/cncf/kubernetes/hooks/kubernetes.py:153]: https://github.com/apache/airflow/pull/9570
Does this problem occur here? What permissions are needed for this method?

File: airflow/providers/cncf/kubernetes/sensors/spark_kubernetes.py [hunk 2]
```diff
@@ -37,38 +39,75 @@ class SparkKubernetesSensor(BaseSensorOperator):
...
 
     @apply_defaults
-    def __init__(self, *,
-                 application_name: str,
-                 namespace: Optional[str] = None,
-                 kubernetes_conn_id: str = 'kubernetes_default',
-                 **kwargs):
+    def __init__(
+        self,
+        application_name: str,
```
Comment (mik-laj) [inline airflow/providers/cncf/kubernetes/sensors/spark_kubernetes.py]:
```suggestion
        self,
        *,
        application_name: str,
```
See: https://github.com/apache/airflow/issues/9942

**Comment by mik-laj** [comment]
Is this log available in another application? Maybe a better solution would be to add a link to another console? https://airflow.readthedocs.io/en/latest/howto/write-logs.html#external-links

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py (hunk 1 above)
Comment (bbenzikry) [reply airflow/providers/cncf/kubernetes/hooks/kubernetes.py:153]: This assumes a role similar to the one below ( specific namespace role or ClusterRole. In any case the operator is constrained to the job namespace ) 

```yaml
- apiGroups: [""]
  resources: ["pods", "pods/log"]
  verbs: ["get", "watch", "list"]
```
It's important to mention that I opted against using the watcher, but left this here for future use ( I'm using ``read_namespaced_pod_log``, which still requires the mentioned permissions )

Should we add a manifest that deals with it for a system test? similar to RBAC yamls from the operator here https://github.com/apache/airflow/blob/7d24b088cd736cfa18f9214e4c9d6ce2d5865f3d/tests/providers/cncf/kubernetes/operators/test_spark_kubernetes_system.py#L36

In any case I don't think it should block, as it depends on explicitly adding the flag

**Comment by bbenzikry** [comment]
> Is this log available in another application? Maybe a better solution would be to add a link to another console? https://airflow.readthedocs.io/en/latest/howto/write-logs.html#external-links

Not necessarily, as it requires either 
A) deploying a spark history server in addition to the operator ( we do have that deployed in our cluster, but it's not applicable for any SparkApplication )
B) setting up log forwarding to ES/equivalent from within the spark context
C) forwarding logs for kubernetes pods to an external system

For a user with airflow access alone, we want the ability to quickly inspect driver logs while inspecting the task without delegating to external systems

**Review by mik-laj — APPROVED** [review]

**Pull request merged**